<a href="https://colab.research.google.com/github/RitikM-AiDev/HuggingFace_Works/blob/main/PHI4_MICROSOFT_TOKENIZER_MODEL.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
phi4 = "microsoft/Phi-4-mini-Instruct"

In [2]:
from transformers import AutoTokenizer

***TOKENIZATION USING MICROSOFT MODEL***

In [4]:
phi4_tokenizer = AutoTokenizer.from_pretrained(phi4)
text = "I am very passionate on learning and I am Genious, I believe In God and I am always Happy and Everyone in My Family are happy always"
tokens = phi4_tokenizer.encode(text)
batch_tokens = phi4_tokenizer.batch_decode(tokens)
tokens

[40,
 939,
 1869,
 28863,
 402,
 7524,
 326,
 357,
 939,
 11820,
 1595,
 11,
 357,
 6423,
 730,
 4809,
 326,
 357,
 939,
 3324,
 27213,
 326,
 29995,
 306,
 3673,
 14098,
 553,
 7150,
 3324]

In [5]:
batch_tokens

['I am very passionate on learning and I am Genious, I believe In God and I am always Happy and Everyone in My Family are happy always']

**CHAT TEMPLATE FOR PHI4 MODEL**

In [9]:
messages = [
    {"role": "system", "content": "You are a helpful assistant."},
    {"role": "user", "content": "What is ML?"}
]

inputs = phi4_tokenizer.apply_chat_template(
    messages,
    return_tensors="pt",
    add_generation_prompt=True
)
inputs

{'input_ids': tensor([[200022,   3575,    553,    261,  10297,  29186,     13, 200020, 200021,
           4827,    382,  52098,     30, 200020, 200019]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

### **What is attention_mask?**

You have:

'attention_mask':
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])

There are 15 values.

And your input_ids also contain 15 token IDs.

So they correspond position-by-position:

input_ids:
[200022, 3575, 553, 261, 10297, ...]

attention:
[     1,  1,   1,  1,     1,     1, ...]

A 1 means:

Pay attention to this token. It is a real part of the input.

A 0 generally means:

Ignore this position. It is padding.